### DISCUSSION TITLE
Missing metrics on services overview page

### DISCUSSION THREAD

**Discussion by CaueP** [SigNoz/signoz#5184] [Q&A] [answered]
Hello 👋🏻 

I noticed that one of my services doesn't have the metrics on the overview dashboard. After some time loading, it shows an error `API responded with 400 - error in builder querie status: error, errors: {}`

Is there something that I can do to fill these charts?

Some screenshots:
Empty charts (loading) on the Overview tab of my service:
![image](https://github.com/SigNoz/signoz/assets/11530788/c0a7cb2e-74ce-4ec6-82ac-f06bb88a4f50)

Error that pops up after a while:
![image](https://github.com/SigNoz/signoz/assets/11530788/6eebc541-8d27-4b23-9b5e-190e9b339955)

The other tabs works fine:
![image](https://github.com/SigNoz/signoz/assets/11530788/1925c189-4abe-4801-9361-7c13f52b2a14)

![Screenshot from 2023-09-06 15-34-57](https://github.com/SigNoz/signoz/assets/11530788/3dd5c470-b624-4389-b4f3-2eb3508a8f46)

## Debugging
Logs that looks related to the error
![image](https://github.com/SigNoz/signoz/assets/11530788/d12b82db-7487-4eca-a170-f0b127a4e749)

Chrome's Network tab with errors
![image](https://github.com/SigNoz/signoz/assets/11530788/e88432d1-a370-4c8e-916c-0e7adcba6bb1)
![image](https://github.com/SigNoz/signoz/assets/11530788/8526fe04-e95f-4f92-933d-4c645eba290b)

**Comment by srikanthccv** [comment, marked as answer]
This page shows the RED metrics based on the entry-point span operations for the service. When such entry-points operations are unusually high this error occurs. The span names (referred to as endpoints) should have low cardinal values however you seem to have some id values in it. Please refer to the OTEL docs and address the high cardinal span names issue https://opentelemetry.io/docs/specs/otel/trace/api/#span.

**Reply by CaueP** [reply]
I didn't even think of it because I'm only using automatic instrumentation.

Thanks a lot, I'll take a look into that.

**Reply by CaueP** [reply]
Do you know if the `name` attribute on the traces explorer is the span name? @srikanthccv 

I made a search on traces explorer grouping by `name`, but it doesn't seem I have a problem with high cardinality.

![image](https://github.com/SigNoz/signoz/assets/11530788/42b70eba-5b3c-4def-a1b1-8d4a5ffce41f)
![image](https://github.com/SigNoz/signoz/assets/11530788/38aa867f-b54d-47fa-b6ab-0ceec042d630)

**Reply by CaueP** [reply]
It's the span name, so it means that the cardinality of this service's spans shouldn't be the problem.

Is there any other thing that I could look into? @srikanthccv

**Reply by srikanthccv** [reply]
> so it means that the cardinality of this service's spans shouldn't be the problem.

By _service's spans_ do you mean the entry-point-spans of a service? SigNoz uses the entry-point-spans referred to as top-level-operation to calculate the RED metrics. For example, in the context of API service, they would be HTTP endpoints. They should be less cardinal.

**Reply by CaueP** [reply]
@srikanthccv  You were right about the issue. I was testing something else and noticed that the spans generated by ASGI lib (opentelemetry-instrumentation-asgi) are adding the ID of my requests to its span name.

Using ASGI: 
![image](https://github.com/SigNoz/signoz/assets/11530788/2422aae1-e653-40b9-be2c-971fbffe5bde)

Using only FastAPI:
![image](https://github.com/SigNoz/signoz/assets/11530788/6dc325b0-8b2b-461c-83b5-d376b68a8adc)


Unfortunately, there is a bug on FastAPI instrumentation that makes the trace context inaccessible when exceptions are thrown (issue https://github.com/open-telemetry/opentelemetry-python/issues/3477), so I rely on ASGI instrumentation to make it work as a workaround. 

Thanks a lot for the support, I'll see how I can handle those noisy spans.

**Reply by CaueP** [reply]
Hey, @srikanthccv. Sorry to bother you again, but I need some help with the old spans.

Even though the retention period is set to 7 days, I have spans that are older than months showing up on the Top Level Operations page. How can I clean up those data or high cardinality spans from Clickhouse DB? 


![image](https://github.com/SigNoz/signoz/assets/11530788/8376d974-1929-4d40-90b2-60586601abc8)

**Reply by CaueP** [reply]
Another screen that we can see that I have only spans from 6/26 and on:

![image](https://github.com/SigNoz/signoz/assets/11530788/64e73340-79e1-4058-b89d-4c8fa486caf6)

**Reply by srikanthccv** [reply]
It will be fixed in https://github.com/SigNoz/signoz/pull/5404 soon. Meanwhile, you can exec into clickhouse and run `DELETE FROM signoz_traces.distributed_top_level_operations ON CLUSTER cluster `. This will remove all the old data and repopulate the table again as the data gets inserted.

**Reply by srikanthccv** [reply]
and attempt to fix the issue in the instrumentation library https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2650

**Reply by CaueP** [reply]
> It will be fixed in #5404 soon. Meanwhile, you can exec into clickhouse and run `DELETE FROM signoz_traces.distributed_top_level_operations ON CLUSTER cluster `. This will remove all the old data and repopulate the table again as the data gets inserted.

Good to know, thanks a lot!



> and attempt to fix the issue in the instrumentation library [open-telemetry/opentelemetry-python-contrib#2650](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2650)

Nice! 🚀

**Reply by CaueP** [reply]
> It will be fixed in #5404 soon. Meanwhile, you can exec into clickhouse and run `DELETE FROM signoz_traces.distributed_top_level_operations ON CLUSTER cluster `. This will remove all the old data and repopulate the table again as the data gets inserted.


Just for the record, this delete operation on `distributed_top_level_operations` table didn't work, probably because it is distributed table as I [found out on this migration](https://github.com/SigNoz/signoz-otel-collector/blob/2b7f110977c4ee60c69766e1899b7eb91face27a/migrationmanager/migrators/traces/migrations/000017_add_distributed_table.up.sql#L16-L17). The table name should be `top_level_operations` instead.

I changed the query to perform the delete on the table `top_level_operations` only for my service with problems, and it worked. The statement used was this:
```
DELETE FROM signoz_traces.top_level_operations WHERE serviceName = 'MY_SERVICE_NAME'
```
